### DISCUSSION TITLE
axmol ios project with CocoaPods

### DISCUSSION THREAD

**Discussion by Dazlonar** [axmolengine/axmol#1721] [Q&A] [answered]
Hello! 
Our iOS project uses [CocoaPods](https://cocoapods.org/) as a dependency management tool. Now I'm trying to migrate our cocos2dx project to axmol, but I'm getting an error during the build with pods.
To rule out that the problem might be in our project, I created a clean axmol project and tried to integrate the pods, but I got the same error:
```
Framework 'Pods_axmoltest' not found
Linker command failed with exit code 1 (use -v to see invocation)
```
I understand that this problem does not directly relate to the axmol engine, but there is probably a problem in the project settings that is causing this to happen. Perhaps someone has an idea why this could be.

- axmol version: 2.1.2
- developing environments
   - Xcode version: 15.2
   - CocoaPods: 1.15.2

Steps to Reproduce:

1. Create a new axmol project with command: `axmol new axmoltest -l cpp -p com.axmol.test -d <your_proj_dir>`
2. Generate XCode proj: `axmol build -p ios -a arm64 -c`
3. [Install CocoaPods](https://guides.cocoapods.org/using/getting-started.html) if needed.
4. Create Podfile for example I am use this: 
```
platform :ios, '12.0'

use_frameworks!
use_modular_headers!

target 'axmoltest' do
  pod 'AppsFlyerFramework'
end
```
5. Install pods dependency: `pod install --repo-update`
6. Open workspace: `axmoltest.xcworkspace`
7. Choose correct target & Run build

**Comment by rh101** [comment]
@Dazlonar Out of curiosity, which version of Cocos2d-x are you migrating from?

**Reply by Dazlonar** [reply]
We have a fairly old project, so this question is quite difficult to answer, because it was a mix of different versions, because no one wanted to completely update before. But this is definitely a version lower than 4. In general, the update to axmol was much easier than we expected. The game looks quite normal, a few game bugs appeared after the update, but I think we can easily fix it. I hope that in a month we will be able to release the first update with axmol. 
Thank you for your work!

**Comment by rh101** [comment]
Just did a test using the steps you provided.  `Pods_axmoltest.framework` does exist, although it is located in a temporary folder (xcode places it there):

`/Users/[user_id]/Library/Developer/Xcode/DerivedData/axmoltest-[somerandomletters]/Build/Products/Debug-iphoneos/Pods_axmoltest.framework`

You can see the location if you click on "Pods_axmoltest" under the Frameworks branch within the Xcode project.

Building it now resulted in the same error: `Framework 'Pods_axmoltest' not found`

So, in order to see if the paths are the issue, I added this to the `CMakeLists.txt`:
`set(CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS "/Users/[user_id]/Library/Developer/Xcode/DerivedData/axmoltest-[somerandomletters]/Build/Products/Debug-iphoneos/")`

Re-generate the xcode project `axmol build -p ios -a arm64 -c`

Re-run `pod install --repo-update`

Open `axmoltest.xcworkspace`, and re-built.  It built successfully.

So, the search path may be the issue.  What I'm not entirely sure about is what you would need to set `CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS` to in order for it to work with any temporary folder that is generated in `/Users/[user_id]/Library/Developer/Xcode/DerivedData`.

**Comment by rh101** [comment]
@Dazlonar Try he following to see if it works for you.

Add this to your `CMakeLists.txt` in the `if(APPLE)` section:

`set(CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS "\${PODS_CONFIGURATION_BUILD_DIR}")`

`PODS_CONFIGURATION_BUILD_DIR` should point to the correct folder.

You also need to add `inherit! :search_paths` the to your Podfile too.

For example:
```
target 'axmoltest' do
  inherit! :search_paths
  pod 'AppsFlyerFramework'
end
```

In my case, the podfile is located in the root folder of the project, and the Xcode project that cmake created is `build_ios_arm64/axmoltest.xcodeproj`, so I have this podfile:

```
platform :ios, '15.0'

use_frameworks!
use_modular_headers!

target 'axmoltest' do
  project './build_ios_arm64/axmoltest.xcodeproj'
  inherit! :search_paths
  pod 'AppsFlyerFramework'
end
```

**Comment by bintester** [comment]
@Dazlonar 
After pod install ,You can try the following steps
1、Open axmoltest.xcworkspace
2、Select target axmoltest and search CONFIGURATION_BUILD_DIR in the Build Settings
3、Modify CONFIGURATION_BUILD_DIR value to  "${BUILD_DIR}/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)",   Ignore quotation marks
4、Clean and rebuild

**Reply by rh101** [reply]
What is the purpose of `CONFIGURATION_BUILD_DIR`? Does it actually affect the search path for the framework?

Also, manually editing the `.xcworkspace` file may not be a good idea, since it's generated outside of the cmake system; isn't it overwritten when you run commands such as `pod install`, meaning you would have to manually edit it every time something changes?

If anything, you should try setting the relevant properties via `CMakeLists.txt`. For example, see if this works:

```
set_target_properties(${target_name} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")
```

**Comment by rh101** [comment, marked as answer]
@Dazlonar Based on what @bintester posted, this also works:

`set(CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")`

I'm not sure if this or the `CONFIGURATION_BUILD_DIR` method is the correct way to deal with this specific issue, but at least there is a solution to it.

**Comment by Dazlonar** [comment]
@rh101 @bintester 
Thank you very much for your answers! 
In the example I provided, all your methods work correctly!
But, if you complicate the Podfile for example like this:
```
platform :ios, '12.2'

use_frameworks!
use_modular_headers!

target 'axmoltest' do
  project './build_ios_arm64/axmoltest.xcodeproj'
  pod 'FirebaseCore'
  pod 'FirebaseAnalytics'
  pod 'FirebaseCrashlytics'
  pod 'FirebaseMessaging'
  pod 'Leanplum-iOS-SDK'
  pod 'nanopb'
  pod 'GoogleToolboxForMac'
  pod 'SendbirdChatSDK'
  pod 'FBSDKCoreKit'
  pod 'FBSDKLoginKit'
  pod 'FBSDKShareKit'
  pod 'AppsFlyerFramework'
  pod 'HelpshiftX', '10.2.3'
  pod 'xcbeautify'

end

```
In this case, `set(CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")` and `set(CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS "\${PODS_CONFIGURATION_BUILD_DIR}")` leads to an error:

```
PhaseScriptExecution [CP]\ Embed\ Pods\ Frameworks /Users/client/Documents/axmoltest/build_ios_arm64/build/axmoltest.build/Debug-iphoneos/Script-74ABECB704E99CE2F877540F.sh (in target 'axmoltest' from project 'axmoltest')
    cd /Users/client/Documents/axmoltest
    /bin/sh -c /Users/client/Documents/axmoltest/build_ios_arm64/build/axmoltest.build/Debug-iphoneos/Script-74ABECB704E99CE2F877540F.sh

mkdir -p /Users/client/Documents/axmoltest/build_ios_arm64/bin/axmoltest/Debug/axmoltest.app/Frameworks
/Users/client/Documents/axmoltest/Pods/Target Support Files/Pods-axmoltest/Pods-axmoltest-frameworks.sh: line 42: source: unbound variable
Command PhaseScriptExecution failed with a nonzero exit code
```
But `set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")` works well for all cases

Thank you very much you helped me a lot!

**Reply by rh101** [reply]
> But `set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")` works well for all cases

Thank you for testing that.  This may be the correct way to handle it.  It's forcing the configuration build directory for the project to match the directory that the pod is using, so all the files will also go into the `/Users/[user_id]/Library/Developer/Xcode/DerivedData/*****` path.

**Reply by rh101** [reply]
@Dazlonar One last thing, the line `project './build_ios_arm64/axmoltest.xcodeproj'` can go at the top of the Podfile.  When I tested this again, but this time using `pod init ./build_ios_arm64/axmoltest.xcodeproj` from the root dir of the project, it automatically added it to the top of the Podfile.
